easybuild/tools/toolchain/compiler.py

```diff
@@ -278,18 +279,40 @@ def _set_optimal_architecture(self, default_optarch=None):
                                 (--optarch and --optarch=GENERIC still override this value)
         """
         optarch = None
+        use_generic = False
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if build_option('optarch') is not None:
+            for optarch_option in build_option('optarch').split(','):
+                # No compilers specified, and option is not generic, so simply take what is there
+                if ":" not in optarch_option and optarch_option != OPTARCH_GENERIC:
+                    optarch = optarch_option
+                    break
+                # Option is generic, so mark it as to be set later and stop iterating
+                elif optarch_option == OPTARCH_GENERIC:
+                    use_generic = True
+                    break
```

Comment:
With this approach, a value like `--optarch=GENERIC,intel:foo` would be equivalent to `--optarch=GENERIC`, while `--optarch=intel:foo,GENERIC` would not be, right? That seems very confusing to me...

I think we should do some validation here, and either except `--optarch=GENERIC`, `--optarch=<flags>` or `--optarch=<flags-by-compiler>`, and not allow mixing; if you detect a dict-like value, we should validate that all entries really specify a mapping, and not make things too flexible (since that makes it too easy to shoot yourself in the foot and end up with a configuration that doesn't do what you think it does).

Another problem I'm concerned with is how to discriminate between `--optarch=<flags>` where the `,` character is included in the list of flags (e.g. `-O2 -Wl,-Bstatic`), and `--optarch=<flags-by-compiler>`, where the entries for the different compiler are separated by a `,`.

So, I think we should go with a more 'obscure' separation character rather than `,`, something like `|` or `@` or `&`, i.e. stuff that is a lot less likely to appear in a set of compiler flags.
The same goes for the `:` character that is not used to separate the compiler name from the set of flags to use for that compiler.

Either that, or let `--optarch` accept the path to a file that specifies to compiler-to-flag mapping in some syntax that's easy to parse (e.g. `.ini` format like we have for external modules metadata, see http://easybuild.readthedocs.io/en/latest/Using_external_modules.html#metadata-for-external-modules).

I actually strongly prefer the latter, since that makes things *a lot* easier; you can just check whether `--optarch` points to an existing file, and fall back to what we do now if it doesn't. Parsing a dict-like value from a string is just painful, and it's probably better to avoid that?